Issue/7540 followed blog notifications

theck13 opened the pull request:
### Fix
This is only ***Part 2 of 2*** for the [Notification for Followed Blogs issue](https://github.com/wordpress-mobile/WordPress-Android/issues/7540).  I divided the issue into two parts so the pull requests aren't too big.

In the Reader, add an item to followed blogs' overflow menu to allow toggling post notifications as well as an actionable message to enable post notifications when a blog is followed.  See the screenshots below for illustration.

![reader_follow](https://user-images.githubusercontent.com/3827611/39450532-e89f33c6-4c88-11e8-8b15-f79c20b38e21.png)

![reader_menu_on](https://user-images.githubusercontent.com/3827611/39450539-ead6c5d2-4c88-11e8-95e8-b297bb4f2392.png)

![reader_menu_off](https://user-images.githubusercontent.com/3827611/39450540-ec639542-4c88-11e8-87a7-939a24d4a248.png)

### Test
#### Follow Button
1. Go to ***Reader*** tab.
2. Tap ***Follow*** button.
3. Notice message and action at bottom of screen.
4. Tap ***Enable*** button.
5. Go to ***Me*** tab.
6. Tap ***Notification Settings*** item.
7. Tap site followed in Step 2 under ***Followed Sites*** section.
8. Notice ***New Posts*** item is enabled.

#### Overflow Menu
1. Go to ***Reader*** tab.
2. Tap overflow menu button of followed site without notifications.
3. Tap ***Turn on site notifications*** item.
4. Go to ***Me*** tab.
5. Tap ***Notification Settings*** item.
6. Tap site followed in Step 2 under ***Followed Sites*** section.
7. Notice ***New Posts*** item is enabled.
8. Go to ***Reader*** tab.
9. Tap overflow menu button of followed site with notifications.
10. Tap ***Turn off site notifications*** item.
11. Go to ***Me*** tab.
12. Tap ***Notification Settings*** item.
13. Tap site followed in Step 9 under ***Followed Sites*** section.
14. Notice ***New Posts*** item is disabled.

folletto commented:
In general looks good.  👍 

Any reason for the snackbar copy to be different?

theck13 commented:
> Any reason for the snackbar copy to be different?

Due to the variable length of site titles and localization translations, it was decided to go with the shorter copy.

The other difference in Design and Develop is the color of the ***Turn off post/site notifications*** menu item.  It was decided green would be better than blue for consistency with the ***Following*** menu item.

folletto commented:
> Due to the variable length of site titles and localization translations, it was decided to go with the shorter copy.

Do you have a link to the discussion? Because this answer puzzled me: I thought the first one being more easily translatable? _--- thinks for a moment ---_ Maybe it's because we can't crop the first line if that gets too long independently from the second line? i.e.:

> Following **Lorem Ipsum Gaium Isciq...**
> Enable post notifications?

Isn't doable?

theck13 commented:
> Do you have a link to the discussion?

The discussion starts [here](https://a8c.slack.com/archives/C8HCZ0CKC/p1524494323000063).

> Maybe it's because we can't crop the first line if that gets too long independently from the second line?

That's correct.  That text is a single view and the lines can't be ellipsized independently.  That was the genesis for changing the copy.

folletto commented:
Understood. Thanks for the clarification – makes sense! 👍

SylvesterWilmott commented:
Looking much better now with dynamic width and those dividers gone!

Can we add the missing 8dp top/bottom paddings on the menu itself per:
https://storage.googleapis.com/material-design/publish/material_v_12/assets/0Bzhp5Z4wHba3aEFjbTF1T1FjQTQ/components-menus-specs1.png

And let's also change the menu background colour to white. Not sure why we use the egg shell colour for these.

theck13 commented:
> Can we add the missing 8dp top/bottom paddings on the menu itself per:
https://storage.googleapis.com/material-design/publish/material_v_12/assets/0Bzhp5Z4wHba3aEFjbTF1T1FjQTQ/components-menus-specs1.png

The style used for the popup menu was change from [Holo](https://github.com/wordpress-mobile/WordPress-Android/pull/7684/files#diff-f192db36c795cee49fabc173724ad0f7L75) (i.e. circa 2007) to [AppCompat](https://github.com/wordpress-mobile/WordPress-Android/pull/7684/files#diff-f192db36c795cee49fabc173724ad0f7R75) (i.e. Material-like compatibility version).

> And let's also change the menu background colour to white. Not sure why we use the egg shell colour for these.

It uses the default/standard [ListPopupWindow](https://github.com/wordpress-mobile/WordPress-Android/pull/7684/files#diff-6c7a325968c55524fd50f1a8ee490c51R1570), which uses `#fafafa` as the background color.  I changed the [style for all dropdown](https://github.com/wordpress-mobile/WordPress-Android/pull/7684/files#diff-f192db36c795cee49fabc173724ad0f7R76) lists to `#ffffff` in 51c6ac0.  See the screenshots below for illustration.

![popup_background](https://user-images.githubusercontent.com/3827611/39560526-fb577ef4-4e5b-11e8-8e91-6b3ab1782449.png)

> Is it also possible to move the menu away from the edge of the screen by 16dp? As it's sitting exactly flush to the edge now.

I can't offset it relative to the screen in a non-hacky way, but I forced the popup view (i.e. action list) gravity to the end of the anchor view (i.e. ellipsis icon) since the anchor view is `16dp` from the edge of the screen, which is effectively the same result, in 4c3bfbf.  See the screenshots below for illustration.

![popup_margin](https://user-images.githubusercontent.com/3827611/39560529-ff6650c4-4e5b-11e8-8914-69a67c87b4be.png)

SylvesterWilmott commented:
Looks great!

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:
```diff
@@ -1558,18 +1584,91 @@ public void onItemClick(AdapterView<?> parent, View view, int position, long id)
+        mDispatcher.dispatch(AccountActionBuilder.newFetchSubscriptionsAction());
+
+        Snackbar.make(view, Html.fromHtml(getString(R.string.reader_followed_blog_notifications,
+                        "<b>", blogName, "</b>")), Snackbar.LENGTH_LONG)
```
When settings a duration of actionable Snackbars, I suggest using AccessibilityUtils.getSnackbarDuration. It designed to provide a longer delay for users who have accessibility services enabled.

theck13 replied:
I didn't know about `AccessibilityUtils.getSnackbarDuration`.  Thanks for the tip!  I updated the durations in 25d0f71.

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/reader/adapters/ReaderMenuAdapter.java:
```diff
@@ -74,7 +77,22 @@ public View getView(int position, View convertView, ViewGroup parent) {
                 textRes = R.string.reader_menu_block_blog;
                 textColorRes = R.color.grey_dark;
-                iconRes = 0;
+                iconRes = R.drawable.ic_block_grey_dark_24dp;
```
Nice!

theck13 replied:
Thanks!

khaykov commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java:
```diff
@@ -416,6 +426,40 @@ public boolean onActivityBackPressed() {
+        mDispatcher.dispatch(AccountActionBuilder.newFetchSubscriptionsAction());
+
+        Snackbar.make(view, Html.fromHtml(getString(R.string.reader_followed_blog_notifications,
+                        "<b>", blogName, "</b>")), Snackbar.LENGTH_LONG)
```
There are some cases when blog name is empty. Should we modify copy to reflect this or leave it like this?

[![https://gyazo.com/b4a23410102059d6d12b55f26a6c2d4b](https://i.gyazo.com/b4a23410102059d6d12b55f26a6c2d4b.png)](https://gyazo.com/b4a23410102059d6d12b55f26a6c2d4b)

theck13 replied:
I updated the copy to say "Enable notifications for this site?" for that case in 65c6da9.  What do you think, @SylvesterWilmott?

theck13 commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java:
```diff
@@ -416,6 +426,40 @@ public boolean onActivityBackPressed() {
+                .setAction(getString(R.string.reader_followed_blog_notifications_action),
+                    new View.OnClickListener() {
+                        @Override public void onClick(View view) {
+                            AddOrDeleteSubscriptionPayload payload = new AddOrDeleteSubscriptionPayload(
```
Yep, we should.  Nice catch!  I added analytics tracking for that in 4f128ef.

theck13 commented:
> I encountered a crash - in some cases, there is a delay between pressing the follow button and Snackbar appearance - if during this delay you navigate away from original activity the crash will happen. To reproduce it, open any post from a reader, press follow button, and close the post details.

I added a check to show it only when the user is still on the Reader detail screen in f783fdd.

> Cash will also occur after orientation change.

I'm unable to get the app to crash on orientation change after tapping the ***Follow*** button on the Reader detail screen.  Can you provide detailed steps to reproduce it?

> when you enable post notifications in Notification Settings, the change is not reflected in reader, but other way around (toggled in reader -> reflected in settings) works ok.

Ahh, I missed that case.  Thanks for pointing it out!  I added the blog table update in 791f3e6.

khaykov commented:
Thanks for the fixes!

The crash with toast on orientation change now only happens when you foollow in the reader list, like this:
[![https://gyazo.com/a501fa694258d649afa9a9be4a7b3266](https://i.gyazo.com/a501fa694258d649afa9a9be4a7b3266.gif)](https://gyazo.com/a501fa694258d649afa9a9be4a7b3266)

```
FATAL EXCEPTION: main
                                     Process: org.wordpress.android, PID: 5413
                                     java.lang.IllegalStateException: Fragment ReaderPostListFragment{96bb47a} not attached to Activity
                                         at android.app.Fragment.getResources(Fragment.java:814)
                                         at android.app.Fragment.getString(Fragment.java:849)
                                         at org.wordpress.android.ui.reader.ReaderPostListFragment.onFollowTapped(ReaderPostListFragment.java:1625)
                                         at org.wordpress.android.ui.reader.views.ReaderSiteHeaderView$3.onActionResult(ReaderSiteHeaderView.java:191)
                                         at org.wordpress.android.ui.reader.actions.ReaderActions.callActionListener(ReaderActions.java:40)
                                         at org.wordpress.android.ui.reader.actions.ReaderBlogActions$1.onResponse(ReaderBlogActions.java:70)
                                         at org.wordpress.android.ui.reader.actions.ReaderBlogActions$1.onResponse(ReaderBlogActions.java:60)
                                         at com.wordpress.rest.RestRequest.deliverResponse(RestRequest.java:76)
                                         at com.wordpress.rest.RestRequest.deliverResponse(RestRequest.java:19)
                                         at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:99)
                                         at android.os.Handler.handleCallback(Handler.java:751)
                                         at android.os.Handler.dispatchMessage(Handler.java:95)
                                         at android.os.Looper.loop(Looper.java:154)
                                         at android.app.ActivityThread.main(ActivityThread.java:6077)
                                         at java.lang.reflect.Method.invoke(Native Method)
                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

 I also found couple more issues:
When you click Enable in the snackbar, the change is not updated in dropdown menu of the posts:
[![https://gyazo.com/b763ae201fbee429d938ef90e5a75e91](https://i.gyazo.com/b763ae201fbee429d938ef90e5a75e91.gif)](https://gyazo.com/b763ae201fbee429d938ef90e5a75e91)



Now, this one is really weird, and I don't have exact steps to reproduce it:
[![https://gyazo.com/d28e26c2f3f0baebc03043e516e611b3](https://i.gyazo.com/d28e26c2f3f0baebc03043e516e611b3.gif)](https://gyazo.com/d28e26c2f3f0baebc03043e516e611b3)

It happens at My Likes screen - sometimes when you follow a blog SnackBar does not appear, and when you press Unfollow it appears at wrong location. 


Also, I somehow managed to get this crash when follwoing and changing filters/going to different screens:
```
FATAL EXCEPTION: main
                                     Process: org.wordpress.android, PID: 6741
                                     java.lang.IllegalArgumentException: No suitable parent found from the given view. Please provide a valid view.
                                         at android.support.design.widget.Snackbar.make(Snackbar.java:137)
                                         at org.wordpress.android.ui.reader.ReaderPostListFragment.onFollowTapped(ReaderPostListFragment.java:1625)
                                         at org.wordpress.android.ui.reader.views.ReaderSiteHeaderView$3.onActionResult(ReaderSiteHeaderView.java:191)
                                         at org.wordpress.android.ui.reader.actions.ReaderActions.callActionListener(ReaderActions.java:40)
                                         at org.wordpress.android.ui.reader.actions.ReaderBlogActions$1.onResponse(ReaderBlogActions.java:70)
                                         at org.wordpress.android.ui.reader.actions.ReaderBlogActions$1.onResponse(ReaderBlogActions.java:60)
                                         at com.wordpress.rest.RestRequest.deliverResponse(RestRequest.java:76)
                                         at com.wordpress.rest.RestRequest.deliverResponse(RestRequest.java:19)
                                         at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:99)
                                         at android.os.Handler.handleCallback(Handler.java:751)
                                         at android.os.Handler.dispatchMessage(Handler.java:95)
                                         at android.os.Looper.loop(Looper.java:154)
                                         at android.app.ActivityThread.main(ActivityThread.java:6077)
                                         at java.lang.reflect.Method.invoke(Native Method)
                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
```

theck13 commented:
> The crash with toast on orientation change now only happens when you foollow in the reader list

Ahh, okay.  I thought you meant it was on the Reader post detail screen.  The crashes are due to the `View` being passed to `Snackbar.make` after the `View` instance has already been removed from the screen (e.g. on device rotation).  To fix that, I updated the follow listener tapped methods to be outside of network calls in 1f80af8.  They will be shown when the Follow button is toggled.  One consequence of that is the snackbar will not reappear after device rotation.  I'm not entirely happy about that behavior, but it's not horrible either.  The message is meant to be ephemeral.  We could make it be shown again after device rotation by added a handful of member variables then saving and restoring them in the instance state bundles, but I don't think that's necessary for a short-lived interface.  What do you think?

> It happens at My Likes screen - sometimes when you follow a blog SnackBar does not appear, and when you press Unfollow it appears at wrong location.

I was never able to reproduce this one.  I'm not even sure how what's happening in that animation is possible.  Let's keep an eye on this and create an issue if we can reproduce it consistently.

> When you click Enable in the snackbar, the change is not updated in dropdown menu of the posts

I added the blog table update for Enable in a2d1f19.
